mage_integrations/mage_integrations/tests/sources/test_base.py

```diff
@@ -0,0 +1,528 @@
+
+    def test_process_stream(self):
+        source = Source()
+        with self.assertRaises(Exception):
+            source.process_stream(
+                CatalogEntry(
+                    replication_method='INVALID_METHOD',
+                    stream='demo_table',
+                    tap_stream_id='demo_table',
+                ),
+            )
+        with patch.object(sys.stdout, 'write') as mock_write:
+            with patch.object(sys.stdout, 'flush') as mock_flush:
+                source.process_stream(
+                    build_sample_streams_catalog().streams[0],
+                )
+                mock_write.assert_called_once()
+                mock_flush.assert_called_once()
+
+    def test_sync_stream(self):
+        source = Source(
+            config=dict(start_date='2023-01-01'),
+        )
+        stream = CatalogEntry(
+            replication_method='FULL_TABLE',
+            stream='demo_users',
+            tap_stream_id='demo_users',
+        )
+        with patch.object(
+            source,
+            '_get_bookmark_properties_for_stream',
+            return_value=['id']
+        ) as mock_get_bookmark_props:
+            with patch.object(source, 'load_data') as mock_load_data:
+                source.sync_stream(stream)
+                mock_get_bookmark_props.assert_called_with(stream)
+                mock_load_data.assert_called_once() 
+
+    def test_write_records(self):
+        source = Source(
+            is_sorted=True,
+        )
+        stream = build_sample_streams_catalog().streams[1]
+        with patch.object(
+            source,
+            '_get_bookmark_properties_for_stream',
+            return_value=['id'],
+        ) as mock_get_bookmark_props:
+            result = source.write_records(
+                stream,
+                [
+                    dict(
+                        age=18,
+                        color='red',
+                        first_name='jason',
+                        id=2,
+                        last_name='scott',
+                    )
+                ],
+            )
+            mock_get_bookmark_props.assert_called_once()
+            self.assertEqual(
+                result,
+                dict(
+                    final_record={
+                        'age': 18,
+                        'id': 2,
+                        'last_name': 'scott',
+                        'first_name': 'jason',
+                        'color': 'red'
+                    },
+                    max_bookmark=[],
+                ),
+            )
+
+    def test_sync(self):
+        catalog = build_sample_streams_catalog()
+        source = Source()
+        with patch.object(source, 'process_stream') as mock_process_stream:
+            with patch.object(source, 'sync_stream') as mock_sync_stream:
+                with patch.object(catalog, 'get_selected_streams', return_value=build_sample_streams_catalog_entries()):
```

Comment:
set maximum line length in your ide. break the line if it exceeds 100 characters.